src/rust/plugin-work-queue/src/psql_queue.rs

```diff
@@ -0,0 +1,305 @@
+pub struct NextExecutionRequest {
+    pub execution_key: ExecutionId,
+    pub plugin_id: uuid::Uuid,
+    pub tenant_id: uuid::Uuid,
+    pub pipeline_message: Vec<u8>,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum PsqlQueueError {
+    #[error("Sqlx")]
+    Sqlx(#[from] sqlx::Error),
+}
+
+#[derive(Clone, Debug)]
+pub struct Message {
+    pub request: NextExecutionRequest,
+}
+
+#[derive(Clone, Debug)]
+pub struct PsqlQueue {
+    pub pool: Pool<Postgres>,
+}
+
+impl PsqlQueue {
+    pub fn new(pool: Pool<Postgres>) -> Self {
+        Self { pool }
+    }
+
+    #[instrument(skip(pipeline_message), err)]
+    pub async fn put_generator_message(
+        &self,
+        plugin_id: Uuid,
+        pipeline_message: Vec<u8>,
+        tenant_id: Uuid,
+    ) -> Result<(), PsqlQueueError> {
+        let now = chrono::Utc::now();
+        sqlx::query!(
+            r"
+            INSERT INTO plugin_work_queue.generator_plugin_executions (
+                plugin_id,
+                pipeline_message,
+                tenant_id,
+                status,
+                creation_time,
+                last_updated,
+                try_count
+            )
+            VALUES( $1::UUID, $2, $3::UUID, 'enqueued', $4, $5, -1 )
+        ",
+            plugin_id,
+            pipeline_message,
+            &tenant_id,
+            &now,
+            now,
+        )
+        .execute(&self.pool)
+        .await?;
+        Ok(())
+    }
+
+    #[instrument(skip(pipeline_message), err)]
+    pub async fn put_analyzer_message(
+        &self,
+        plugin_id: Uuid,
+        pipeline_message: Vec<u8>,
+        tenant_id: Uuid,
+    ) -> Result<(), PsqlQueueError> {
+        let now = chrono::Utc::now();
+        sqlx::query!(
+            r"
+            INSERT INTO plugin_work_queue.analyzer_plugin_executions (
+                plugin_id,
+                pipeline_message,
+                tenant_id,
+                status,
+                creation_time,
+                last_updated,
+                try_count
+            )
+            VALUES( $1::UUID, $2, $3::UUID, 'enqueued', $4, $5, -1 )
+        ",
+            plugin_id,
+            pipeline_message,
+            &tenant_id,
+            &now,
+            now,
+        )
+        .execute(&self.pool)
+        .await?;
+        Ok(())
+    }
+
+    #[instrument(err)]
+    pub async fn get_generator_message(&self) -> Result<Option<Message>, PsqlQueueError> {
+        // `get_message` does a few things
+        // 1. It attempts to get a message from the queue
+        //      -> Where that message isn't over a day old
+        //      -> Where that message is "visible"
+        //      -> Where that message isn't currently being evaluated by another transaction
+        //      -> Where that message is in the 'enqueued' state
+        // 2. Updates the `try_count`
+        // 3. Updates the `visible_after`
+
+        // Note that:
+        // * messages are invisible for 10 seconds
```

Comment:
Can you say more about the intention of "invisibility" here? Is it effectively saying that this process has a "lock" on this record for the next little bit to ensure it's not being used elsewhere?

If so, we could probably simplify this a bit with row locks, as described in https://spin.atomicobject.com/2021/02/04/redis-postgresql.